### PR TITLE
feat: allow entering non-default values in multi-select

### LINKED ISSUE #15488: cliui: Allow entering non-default values in MultiSelect
# Problem

`cliui.Multiselect` currently does not allow users to input custom options if zero choices are provided.

## Use case

As a template author, I want to allow my users to allow setting arbitrary environment variables inside their workspace container. 

I provide a parameter similar to the below:
```
data "coder_parameter" "env_vars" {
  name         = "env_vars"
  display_name = "Additional environment variables"
  type         = "list(string)"
  default      = jsonencode([])
  description  = "key=value pairs to set in container"
  mutable      = true
}
```

## What I want

My users should then be able to specify additional environment variables when creating workspaces via both UI and CLI:

```
Additional environment variables
  key=value pairs to set in workspace environment
> foo=bar
> baz=zap
?    [Type to add an option, or press Enter to finish]
Planning workspace...
```

## What I get

Users can specify additional environment variables via UI, but CLI does not work:

```
Additional environment variables
  key=value pairs to set in workspace environment <-- does not even prompt the user to add more options
Planning workspace...
```

The only workaround for CLI is to specify parameters via `--rich-parameter-file`. 

# Proposed Solution

`cliui.MultiSelect` should always allow users to specify another option.
For example, when running `coder exp prompt-example multi-select`:

```
? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
> [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  ... Other

# Input: 🔽🔽🔽🔽🔽

? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
  [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
>  ... Other

# Input: ⏎

? Type what you want: [Enter to submit]: 
>  

# Input: potato⏎

? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
  [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  [x] potato
>  ... Other

```

### PR DESCRIPTION
**Pull request by joobisb**
Closes https://github.com/coder/coder/issues/15488

cliui.Multiselect currently does not allow users to input custom options if zero choices are provided.

### Proposed Solution

cliui.MultiSelect should always allow users to specify another option.

For example, when running coder exp prompt-example multi-select: 

PS: I have modified the text a bit from the original issue

```
? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
> [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  ... Other
+ Add custom value


? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
? Enter custom value: potato⏎


? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
  [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  [x] potato
>  ... Other
+ Add custom value

```

Attaching a recording for the same

https://github.com/user-attachments/assets/ce02cfab-4549-4431-b0d7-c7c2d69ae7ca

### CODE REVIEW

File: cli/cliui/select_test.go [hunk 1]
```diff
@@ -101,6 +101,39 @@ func TestMultiSelect(t *testing.T) {
 		}()
 		require.Equal(t, items, <-msgChan)
 	})
+
+	t.Run("MultiSelectWithCustomInput", func(t *testing.T) {
```
Comment (joobisb) [inline cli/cliui/select_test.go:105]: couldn't add tests to check the custom input flow, we had this check here and not sure if its safe to remove and add interactive tests, please do let me know if there's a better way to incorporate those tests

https://github.com/coder/coder/blob/63572d9f5370a860653adf9ff61256a7aed1724c/cli/cliui/select.go#L310

**Comment by joobisb** [comment]
Hi @johnstcn ,

could you please have a look at this PR?

I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?

**Comment by johnstcn** [comment]
> I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?

I think it would make more sense to either de-duplicate or validate that the custom option isn't already present.

File: cli/cliui/select_test.go (hunk 1 above)
Comment (johnstcn) [reply cli/cliui/select_test.go:105]: I think historically we've manually tested this using `prompt-example`. I agree that it's not ideal :-(

File: cli/cliui/select.go [hunk 2]
```diff
@@ -370,12 +372,15 @@ type multiSelectOption struct {
...
-	message  string
-	canceled bool
-	selected bool
+	search            textinput.Model
+	options           []*multiSelectOption
+	cursor            int
+	message           string
+	canceled          bool
+	selected          bool
+	isInputMode       bool   // New field to track if we're adding a custom option
```
Comment (johnstcn) [inline cli/cliui/select.go:381]: suggestion: `isCustomInputMode` for consistency

File: cli/cliui/select.go [hunk 3]
```diff
@@ -414,15 +428,17 @@ func (m multiSelectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 
 		case tea.KeyUp:
 			options := m.filteredOptions()
+			maxIndex := len(options)
 			if m.cursor > 0 {
 				m.cursor--
 			} else {
-				m.cursor = len(options) - 1
+				m.cursor = maxIndex
```
Comment (DanielleMaywood) [inline cli/cliui/select.go:435]: I think this logic needs to handle when `EnableCustomInput` is false. Currently it lets you go over the last valid option.

File: cli/prompts.go [hunk 4]
```diff
@@ -156,9 +156,10 @@ func (RootCmd) promptExample() *serpent.Command {
 					multiSelectValues, multiSelectError = cliui.MultiSelect(inv, cliui.MultiSelectOptions{
 						Message: "Select some things:",
 						Options: []string{
-							"Code", "Chair", "Whale", "Diamond", "Carrot",
+							"Code", "Chairs", "Whale", "Diamond", "Carrot",
 						},
-						Defaults: []string{"Code"},
+						Defaults:          []string{"Code"},
+						EnableCustomInput: true,
```
Comment (johnstcn) [inline cli/prompts.go:162]: Can we expose this as a flag so we can easily test both with and without custom input?

**Comment by joobisb** [comment]
> > I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?
> 
> I think it would make more sense to either de-duplicate or validate that the custom option isn't already present.

@johnstcn @DanielleMaywood 

I've addressed the comments and handled duplicates as shown in the recording

- case insensitive comparison - if it exists, it selects and cursor points to the already existing option
- if not add a new option

https://github.com/user-attachments/assets/28d75728-b392-48cc-9a1d-2048c2ab3363

File: cli/cliui/select.go [hunk 5]
```diff
@@ -457,6 +471,91 @@ func (m multiSelectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
...
+		m.isCustomInputMode = false
+		return m, nil
+	}
+
+	// Clear search to ensure option is visible and cursor points to the new option
+	m.search.SetValue("")
+
+	// Check for duplicates
+	for i, opt := range m.options {
+		if strings.EqualFold(opt.option, m.customInput) {
```
Comment (johnstcn) [inline cli/cliui/select.go:520]: I'm not 100% sure about the case-insensitive matching. On the surface, it seems innocuous enough but I'm not sure if there's going to be a case where someone is going to want to differentiate `A` from `a`. Let's keep matching case-sensitive for now.
We can add an option for case-insensitive matching later if someone requests it.

**Review by johnstcn — APPROVED** [review]

File: cli/cliui/select.go (hunk 5 above)
Comment (joobisb) [reply cli/cliui/select.go]: updated the PR!

**Pull request merged**